File: hello-csharp/src/Server/Message.cs
using System.Text.Json;
using System.Text.Json.Serialization;

namespace AudioStreamServer;

/// <summary>JSON message DTO for WebSocket communication.</summary>
public class Message
{
    [JsonPropertyName("command")]
    public string Command { get; set; } = "";

    [JsonPropertyName("streamId")]
    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public string? StreamId { get; set; }

    [JsonPropertyName("offset")]
    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public long? Offset { get; set; }

    [JsonPropertyName("length")]
    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public int? Length { get; set; }

    [JsonPropertyName("message")]
    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public string? Msg { get; set; }

    [JsonPropertyName("status")]
    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public string? Status { get; set; }

    [JsonPropertyName("size")]
    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public long? Size { get; set; }

    [JsonPropertyName("streams")]
    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public string? Streams { get; set; }

    // ── Factory methods ─────────────────────────────────────────────────

    public static Message Connected(string connId) => new() { Command = "CONNECTED", StreamId = connId };
    public static Message Created(string streamId) => new() { Command = "CREATED", StreamId = streamId };
    public static Message Completed(string streamId) => new() { Command = "COMPLETED", StreamId = streamId };
    public static Message Closed(string streamId) => new() { Command = "CLOSED", StreamId = streamId };
    public static Message Error(string message) => new() { Command = "ERROR", Msg = message };

    public static Message StreamStatus(string streamId, string status, long size) =>
        new() { Command = "STATUS", StreamId = streamId, Status = status, Size = size };

    public static Message StreamList(IEnumerable<string> ids) =>
        new() { Command = "STREAM_LIST", Streams = string.Join(",", ids) };

    // ── Parse ───────────────────────────────────────────────────────────

    private static readonly JsonSerializerOptions JsonOpts = new()
    {
        PropertyNameCaseInsensitive = true,
    };

    public static (Message msg, CommandInfo info) Parse(string json)
    {
        var m = JsonSerializer.Deserialize<Message>(json, JsonOpts)
                ?? throw new ArgumentException("Invalid JSON");
        var info = CommandInfo.Lookup(m.Command);
        return (m, info);
    }

    public string ToJson() => JsonSerializer.Serialize(this, JsonOpts);
}

File: hello-kotlin/src/main/kotlin/server/Message.kt
package server

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json

val json = Json { ignoreUnknownKeys = true; encodeDefaults = false }

/** WebSocket JSON message DTO. */
@Serializable
data class Message(
    @SerialName("command") val command: String,
    @SerialName("streamId") val streamId: String? = null,
    @SerialName("offset") val offset: Long? = null,
    @SerialName("length") val length: Int? = null,
    @SerialName("message") val message: String? = null,
    @SerialName("status") val status: String? = null,
    @SerialName("size") val size: Long? = null,
    @SerialName("streams") val streams: String? = null,
) {
    fun toJson(): String = json.encodeToString(serializer(), this)

    companion object {
        fun connected(connId: String)  = Message(command = "CONNECTED", streamId = connId)
        fun created(streamId: String)  = Message(command = "CREATED", streamId = streamId)
        fun completed(streamId: String)= Message(command = "COMPLETED", streamId = streamId)
        fun closed(streamId: String)   = Message(command = "CLOSED", streamId = streamId)
        fun error(msg: String)         = Message(command = "ERROR", message = msg)

        fun status(streamId: String, status: String, size: Long) =
            Message(command = "STATUS", streamId = streamId, status = status, size = size)

        fun streamList(ids: List<String>) =
            Message(command = "STREAM_LIST", streams = ids.joinToString(","))

        fun parse(text: String): Pair<Message, CommandInfo> {
            val m = json.decodeFromString(serializer(), text)
            val info = CommandInfo.lookup(m.command)
            return m to info
        }
    }
}
